### PR TITLE
Check init arguments for missing dunder notation

### LINKED ISSUE #500: Catching wrong parameters passed to NeuralNet
Right now, NeuralNet et al. are very forgiving in accepting parameters. Part of that has to do with the dynamic nature of skorch but some typos can probably be caught. E.g. in #498, there was the issue that a user passed `iterator_train_num_workers=4` (used single instead of double underscore), which we should be able to identify.

### PR DESCRIPTION
**Pull request by BenjaminBossan**
Fixes #500 

This catches wrong keyword arguments such as `optimizerlr` or `iterator_train_shuffle` and suggests the correct name.

### CODE REVIEW

File: skorch/net.py [hunk 1]
```diff
@@ -229,20 +229,7 @@ def __init__(
...
-            if any(key.startswith(p) for p in self.prefixes_):
-                continue
-            unexpected_kwargs.append(key)
-        if unexpected_kwargs:
-            msg = ("__init__() got unexpected argument(s) {}. "
-                   "Either you made a typo, or you added new arguments "
-                   "in a subclass; if that is the case, the subclass "
-                   "should deal with the new arguments explicitely.")
-            raise TypeError(msg.format(', '.join(sorted(unexpected_kwargs))))
+        kwargs = self.check_kwargs(kwargs)
```
Comment (thomasjpfan) [inline skorch/net.py]: Private for now? I do not think users would want to override this.

File: skorch/net.py [hunk 2]
```diff
@@ -1285,6 +1272,55 @@ def get_params(self, deep=True, **kwargs):
         params.update(params_cb)
         return params
 
+    def check_kwargs(self, kwargs):
+        """Check argument names passed at initialization.
```
Comment (thomasjpfan) [inline skorch/net.py]: For our future selves, maybe add a small example on how `iterator_train_shuffle` will raise a `TypeError` that will suggest `iterator_train__shuffle`.

File: skorch/net.py [hunk 3]
```diff
@@ -1285,6 +1272,55 @@ def get_params(self, deep=True, **kwargs):
...
+        missing_dunder_kwargs = []
+        for key in kwargs:
+            if key.endswith('_'):
+                continue
+            for prefix in self.prefixes_:
+                if key.startswith(prefix):
+                    if not key.startswith(prefix + '__'):
+                        missing_dunder_kwargs.append((prefix, key))
+                    break
+            else:  # if no break
```
Comment (thomasjpfan) [inline skorch/net.py]: Got to love this very obvious python feature.

File: skorch/tests/test_net.py [hunk 4]
```diff
@@ -178,6 +178,48 @@ def test_net_init_two_unknown_argument(self, net_cls, module_cls):
                     "should deal with the new arguments explicitely.")
         assert e.value.args[0] == expected
 
+    def test_net_init_missing_dunder_in_prefix_argument(
```
Comment (thomasjpfan) [inline skorch/tests/test_net.py]: We can `pytest.mark.parametrize` and test for some other prefixes like `optimizer_momentum`:

```py
@pytest.mark.parametrize("key, value", 
        [("iterator_train_shuffle", True), ("optimizer_momentum", 0.95)]
```

**Comment by BenjaminBossan** [comment]
@thomasjpfan Thx for the feedback, I worked that in.

As an aside: I just noticed that this test now passes even though it should xfail:

https://github.com/skorch-dev/skorch/blob/f94466e272f6f325898359fecb9a7c004354af7f/skorch/tests/test_net.py#L1155-L1167

Not sure what change caused this.

**Comment by thomasjpfan** [comment]
The `test_get_params_with_uninit_callbacks` test passes on master as well.

File: CHANGES.md [hunk 5]
```diff
@@ -9,6 +9,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
+- More careful check for wrong parameter names being passed to NeuralNet
```
Comment (thomasjpfan) [inline CHANGES.md]: Reference this PR or issue. (I have found reference the PR is more useful, since it usually also links to the issue)

Comment (BenjaminBossan) [reply CHANGES.md]: True. Added it.

**Comment by BenjaminBossan** [comment]
> The `test_get_params_with_uninit_callbacks` test passes on master as well.

Yes, now it's hard to find out what change caused this except for bisecting. We may want to add `pytest.mark.xfail(strict=True)` in the future.

File: skorch/net.py (hunk 3 above)
Comment (ottonemo) [reply skorch/net.py]:
```suggestion
            else:  # no break means key didn't match a prefix
```

**Pull request merged**